81-IDevID_and_IAK_support.md

```diff
@@ -0,0 +1,280 @@
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a enhancement and for
+highlighting any additional information provided beyond the standard enhancement
+template.
+-->
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [keylime/enhancements] referencing this enhancement and targeting a release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Keylime i.e., [keylime/keylime], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to pull request in [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementers to split their attention between writing release
+notes and implementing the feature itself. Reviewers
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+This enhancement proposes leveraging the IDevID and IAK keys/certificates to register each device at the registrar service. This can be used for later enabling the generation and use of LDevID and LAKs, as can also allow attestion based on the IAK instead of using an ephemeral AK (Attestation key) if desired. IDevID is an industry standard identity which is issued by manufacturers when the device is built.
```

Comment:
Can you give an introduction about some of the concepts? Seems that you used the "TPM 2.0 Keys for Device
Identity and Attestation" document for this paragraph, where also the ldevid is described.

[1] https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_DevID_v1r2_02dec2020.pdf

IIUC the idevid should be stored in 0x81020000, what to do if it is missing? Also I fail to see why idevid is more relevant that the AK. Is it an advantage that never changes, or there are more reasons?

Reply:
Thanks for the feedback. I just updated the PR with more info that we have been discussing at the slack channel at [1] during the last week or so. Please check the thread if possible.

Trying to summarize on the points you mentioned, the document "TPM 2.0 Keys for Device Identity and Attestation" [2] specifies the concept and differences between the different TPM keys available. While the EK has information about the TPM manufacturer, the IDevID/IAK (“I” for initial) are generated by the device OEM/manufacturer. 
They are credentials intended to be usable for the life of the product. The IDevID can be used to assure the device is authentic by having info on the Serial Number and SKU and has certificates generated/signed by the OEM. As an example, HPE Aruba has several million devices in the field with IDevID certificate issued to TPMs in order to give assurance the device is authentic. The AK cannot provide that level of assurance bound to the physical device. 
In case the IDevID key is missing, it can be regenerated based on a template. But as pointed on the updated PR, we included 4 different use cases, and in case the device does not have an IDevID and IAK, the default and current scenario with EK + AK can be used.

Finally, we have the possibility of generation and use of LDevIDs and LAKs by the product client/owner, which can be based on the Initial keys provided by the OEM. The figure at the item "5.7 Summary of EK, IDevID/IAK and LDevID/LAK Relationships" at [2] describes the relation between the different keys.
 
1. https://cloud-native.slack.com/archives/C01ARE2QUTZ/p1658936450005719
2. https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_DevID_v1r2_02dec2020.pdf